cmd/kubelet/app/server.go

```diff
@@ -440,6 +449,131 @@ func InitializeTLS(s *options.KubeletServer) (*server.TLSOptions, error) {
+//     The private key is stored to disk.
+// (3) Send request to API server and watch for the issued certificate.
+// (4) Once (3) succeeds, dump the certificate to disk.
+func requestCertFromAPIServer(s *options.KubeletServer) error {
+	// (1).
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err != nil {
+		return fmt.Errorf("unable to create API server client config: %v", err)
+	}
+
+	certificatesclient, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to create certificates client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
+	if nodeIP != nil {
+		ips = []net.IP{nodeIP}
+	}
+
+	req, err := utilcertificates.NewCertificateRequest(s.TLSPrivateKeyFile, &pkix.Name{CommonName: hostname}, []string{hostname}, ips)
+	if err != nil {
+		return fmt.Errorf("unable to generate certificate request: %v", err)
+	}
+
+	// (3).
+	certificate, err := requestCertificate(certificatesclient, req, 3600) // Make a default timeout = 3600s
+	if err != nil {
+		return fmt.Errorf("unable to request certificate from API server: %v", err)
+	}
+
+	// (4).
+	if err := crypto.WriteCertToPath(s.TLSCertFile, certificate); err != nil {
+		return fmt.Errorf("unable to create certificate file on disk: %v", err)
+	}
+
+	return nil
+}
+
+// requestCertificate sends the certificate signing request to API server and watches the object's status.
+// It returns the API server's issued certificate (pem-encoded) on success.
+// If there is any errors, or the watch timeouts, it returns an error.
+func requestCertificate(client unversionedcertificates.CertificateSigningRequestsGetter, request []byte, defaultTimeoutSeconds int64) (certificate []byte, err error) {
+	res, err := client.CertificateSigningRequests().Create(&certificates.CertificateSigningRequest{
+		TypeMeta:   unversioned.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: api.ObjectMeta{GenerateName: "csr-"},
+
+		// Username, UID, Groups will be injected by API server.
+		Spec: certificates.CertificateSigningRequestSpec{Request: request},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot create certificate signing request: %v", err)
+	}
+
+	defer client.CertificateSigningRequests().Delete(res.Name, nil)
+
+	resultCh, err := client.CertificateSigningRequests().Watch(api.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &defaultTimeoutSeconds,
+		// Label and field selector are not used now.
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot watch on the certificate signing request: %v", err)
+	}
+
+	var status certificates.CertificateSigningRequestStatus
+	ch := resultCh.ResultChan()
+
+	for {
+		event, ok := <-ch
+		if !ok {
+			break
+		}
+
+		if event.Type == watch.Modified {
+			status = event.Object.(*certificates.CertificateSigningRequest).Status
+			for _, c := range status.Conditions {
```

Comment:
> But I think we still need to make sure we have "approved" condition before looking the cert.

Sure. I just wanted to get something working. I really am not sure how the condition ends up not being set.